### PR TITLE
Improve command system; add aliases for commands

### PR DESCRIPTION
**Pull request by lengweiping1983**
> **Note** This PR has been "hijacked" and expanded by a maintainer to further increase its value and accelerate its integration.

### Background
The AI will regularly hallucinate command names. By providing probable synonyms (aliases) for existing commands, we can prevent failure in such cases.

Before:
```py
@command(
    "write_to_file",
    "Write to file",
    '"filename": "<filename>", "text": "<text>"',
)
```

Now:
```py
@command(
    "write_to_file",
    "Write to file",
    '"filename": "<filename>", "text": "<text>"',
    name_alias=["write_file", "create_file"],
)
```
`write_to_file`, `write_file`, and `create_file` will all take effect.

### Changes
* `CommandRegistry`
  * Add `__contains__()` method to support checking `command_name in registry`
  * Use `commands_aliases` to map aliases to commands in `get_command()`, `call()` and `__contains__()`
  * `unregister()` now takes a `Command` as its argument instead of `str`
  * `get_command()` now returns `None` if a command does not exist, instead of a `KeyError`
* `PromptGenerator`
  * Fix import error for `CommandRegistry`
  * Add type hints to attributes
  * `args` -> `params`
* `app.py::execute_command()`
  * Use `CommandRegistry.get_command()` instead of `CommandRegistry.commands.get()`
  * Improve comments to clarify code flow and purpose
* Rename `app.py::get_command()` to `extract_command` to clarify its function
* test_prompt_generator.py
  * Update tests in accordance with changes to `PromptGenerator`
  * Delete exact duplicate test `test_generate_prompt_string`
* test_commands.py
  * Convert to pytest-style
    * Add fixtures to reduce duplicate code
  * Reduce overlap between some tests
  * Use `command_name in registry` instead of `command_name in registry.commands`
  * Add `test_command_in_registry` to test `in` operator

### Documentation
x

### Test Plan
CI; see changes to tests

### PR Quality Checklist
- [ ] My pull request is atomic and focuses on a single change.
- [x] I have thoroughly tested my changes with multiple different prompts.
- [x] I have considered potential risks and mitigations for my changes.
- [x] I have documented my changes clearly and comprehensively.
- [ ] I have not snuck in any "extra" small tweaks changes

### CODE REVIEW

**Comment by lengweiping1983** [comment]
@nponeccop @BillSchumacher

this PR Extended Command Features.
Command Multiple names are supported, in the form of aliases. 

please review.  https://github.com/Significant-Gravitas/Auto-GPT/pull/2635

**Comment by lengweiping1983** [comment]
@Pwuts  @nponeccop  Today I found I used the company email to submit the PR, now I changed to my personal email and resubmitted.

**Comment by lengweiping1983** [comment]
@BillSchumacher  please review this PR, and https://github.com/Significant-Gravitas/Auto-GPT/pull/2311

**Comment by p-i-** [comment]
This is a mass message from the AutoGPT core team.
Our apologies for the ongoing delay in processing PRs.
This is because we are re-architecting the AutoGPT core!

For more details (and for infor on joining our Discord), please refer to:
  https://github.com/Significant-Gravitas/Auto-GPT/wiki/Architecting

File: autogpt/commands/command.py [hunk 1]
```diff
@@ -26,13 +26,20 @@ def __init__(
         signature: str = "",
         enabled: bool = True,
         disabled_reason: Optional[str] = None,
+        name_alias: Optional[Union[str, list]] = None,
```
Comment (lc0rp) [inline autogpt/commands/command.py]: "alias" or "aliases" seems more concise than name_alias.

Comment (lc0rp) [reply autogpt/commands/command.py]: I'd also make it `Optional[list]`

File: autogpt/commands/command.py [hunk 2]
```diff
@@ -66,30 +73,39 @@ def register(self, cmd: Command) -> None:
                 f"Command '{cmd.name}' already registered and will be overwritten!"
             )
         self.commands[cmd.name] = cmd
+        for command_name_alias in cmd.name_alias:
+            if command_name_alias == cmd.name:
+                continue
+            self.commands[command_name_alias] = cmd
```
Comment (lc0rp) [inline autogpt/commands/command.py]: - Each alias will be sent as a separate command to the LLM, duplicating descriptions and signatures. This results in increased use of the context window and could confuse the LLM. Can we include aliases in the same command output sent to the LLM?
- I'm curious how this works with the new Open AI function calling feature.

File: autogpt/commands/command.py
Comment (lc0rp) [inline autogpt/commands/command.py:1]: I would combine this with the synonym feature, either replacing it or expanding it.

**Review by lc0rp — COMMENTED** [review]
- Interesting feature. It needs benchmarks to ensure it doesn't confuse the LLM and eat up too much of the context window. Also, the synonyms feature should be taken into account.
- Lastly, some variable renames weren't necessary and/or related to this feature

File: autogpt/commands/command.py
Comment (Pwuts) [reply autogpt/commands/command.py:1]: Done

File: autogpt/commands/command.py (hunk 2 above)
Comment (Pwuts) [reply autogpt/commands/command.py]: Reverted this; I think duplicating the command definition in the prompt won't add much value. It's more about catching hallucinations.

File: autogpt/commands/command.py (hunk 1 above)
Comment (Pwuts) [reply autogpt/commands/command.py]: `list[str] = []` works too :)

**Review by merwanehamadi — APPROVED** [review]

**Review by merwanehamadi — APPROVED** [review]

**Pull request merged**